File: backend/src/main/java/com/backend/api/user/service/EmailService.java
package com.backend.api.user.service;

import com.backend.domain.user.entity.AccountStatus;
import com.backend.domain.user.entity.User;
import com.backend.domain.user.repository.VerificationCodeRepository;
import com.backend.domain.userPenalty.entity.UserPenalty;
import com.backend.global.exception.ErrorCode;
import com.backend.global.exception.ErrorException;
import com.backend.domain.user.entity.VerificationCode;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.mail.SimpleMailMessage;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.LocalDateTime;

@Slf4j
@Service
@RequiredArgsConstructor
public class EmailService {

    private final JavaMailSender mailSender;
    private final VerificationCodeRepository verificationCodeRepository;

    private static final String CHAR_SET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    private static final int CODE_LENGTH = 6;


    @Transactional
    public void createAndSendVerificationCode(String email) {
        verificationCodeRepository.findByEmail(email)
                .ifPresent(verificationCodeRepository::delete);

        String code = generateVerificationCode();

        VerificationCode verification = VerificationCode.builder()
                .email(email)
                .code(code)
                .expiresAt(LocalDateTime.now().plusMinutes(5))
                .verified(false)
                .build();

        verificationCodeRepository.save(verification);

        // 비동기 메일 전송
        sendVerificationMailAsync(email, code);
    }

    @Async("mailExecutor")
    void sendVerificationMailAsync(String email, String code) {
        try {
            SimpleMailMessage message = new SimpleMailMessage();
            message.setTo(email);
            message.setSubject("[Dev-Station] 이메일 인증 코드");
            message.setText("""
                안녕하세요. Dev-Station 입니다.
                
                회원가입을 위해 아래 인증 코드를 입력해주세요.
                인증코드: %s
                
                해당 코드는 5분간 유효합니다.
                """.formatted(code));
            mailSender.send(message);
            log.info("[이메일 인증] 인증코드 전송 완료: {}", email);
        } catch (Exception e) {
            log.error("이메일 인증코드 전송 실패: {}", e.getMessage(), e);
            throw new ErrorException(ErrorCode.EMAIL_SEND_FAILED);
        }
    }

    private String generateVerificationCode() {
        SecureRandom random = new SecureRandom();
        StringBuilder sb = new StringBuilder(CODE_LENGTH);
        for (int i = 0; i < CODE_LENGTH; i++) {
            int index = random.nextInt(CHAR_SET.length());
            sb.append(CHAR_SET.charAt(index));
        }
        return sb.toString();
    }

    // 이메일 인증 코드 검증
    public void verifyCode(String email, String code) {
        VerificationCode verification = verificationCodeRepository.findByEmail(email)
                .orElseThrow(() -> new ErrorException(ErrorCode.INVALID_VERIFICATION_CODE));

        if (verification.isExpired()) {
            throw new ErrorException(ErrorCode.EXPIRED_VERIFICATION_CODE);
        }

        if (!verification.getCode().equals(code)) {
            throw new ErrorException(ErrorCode.INVALID_VERIFICATION_CODE);
        }

        verification.markAsVerified();
        verificationCodeRepository.save(verification);
        log.info("[이메일 인증] 인증 성공: {}", email);
    }

    // 인증 여부 확인
    public boolean isVerified(String email) {
        return verificationCodeRepository.findByEmail(email)
                .map(VerificationCode::isVerified)
                .orElse(false);
    }

    // 회원가입 완료 시 환영 메일 발송
    @Async("mailExecutor")
    public void sendWelcomeMail(User user) {
        try {
            SimpleMailMessage message = new SimpleMailMessage();
            message.setTo(user.getEmail());
            message.setSubject("[Dev-Station] 회원가입을 환영합니다!");
            message.setText("""
                안녕하세요, %s님 👋
                
                Dev-Station에 가입해 주셔서 감사합니다.
                지금부터 CS 인터뷰 문제 풀이, AI 피드백, 프로젝트 모집 등 모든 기능을 이용하실 수 있습니다.
                
                앞으로도 좋은 서비스로 보답하겠습니다!
                
                - Dev-Station 팀 드림 -
                """.formatted(user.getName()));

            mailSender.send(message);
            log.info("[회원가입 메일] 환영 메일 전송 완료: {}", user.getEmail());
        } catch (Exception e) {
            log.error("회원가입 환영 메일 전송 실패: {}", e.getMessage(), e);
        }
    }

    // 계정 상태 변경 메일 (정지 / 복구 / 탈퇴 등)
    @Async("mailExecutor")
    public void sendStatusChangeMail(User user, UserPenalty penalty) {
        try {
            AccountStatus status = user.getAccountStatus();
            String subject;
            String content;

            switch (status) {
                case SUSPENDED -> {
                    subject = "[Dev-Station] 계정 일시정지 안내";
                    content = """
                            안녕하세요, %s님.
                            
                            회원님의 계정이 현재 '일시정지' 상태로 전환되었습니다.
                            
                            📌 정지 사유: %s
                            📅 정지 해제 예정일: %s
                            
                            정책 위반 혹은 신고 누적으로 인한 조치일 수 있습니다.
                            자세한 내용은 관리자에게 문의 바랍니다.
                            
                            문의: support@devstation.com
                            """.formatted(
                            user.getName(),
                            penalty.getReason(),
                            penalty.getEndAt() != null ? penalty.getEndAt().toLocalDate() : "미정"
                    );
                }
                case BANNED -> {
                    subject = "[Dev-Station] 계정 영구 정지 안내";
                    content = """
                            안녕하세요, %s님.
                            
                            회원님의 계정이 '영구 정지' 처리되었습니다.
                            
                            📌 정지 사유: %s
                            
                            중대한 정책 위반으로 인해 재가입이 제한됩니다.
                            문의가 필요하신 경우 support@devstation.com 으로 연락 바랍니다.
                            """.formatted(
                            user.getName(),
                            penalty.getReason()
                    );
                }
                case ACTIVE -> {
                    subject = "[Dev-Station] 계정 복구 안내";
                    content = """
                            안녕하세요, %s님.
                            
                            회원님의 계정이 정상 상태로 복구되었습니다.
                            지금부터 정상적으로 로그인 및 활동이 가능합니다.
                            """.formatted(user.getName());
                }
                case DEACTIVATED -> {
                    subject = "[Dev-Station] 회원 탈퇴 완료 안내";
                    content = """
                            안녕하세요, %s님.
                            
                            회원님의 계정 탈퇴 처리가 완료되었습니다.
                            탈퇴 이후에도 일정 기간 동안 개인정보가 보관될 수 있습니다.
                            다시 서비스를 이용하시려면 재가입을 진행해주세요.
                            """.formatted(user.getName());
                }
                default -> {
                    log.info("이메일 전송 대상이 아닌 상태: {}", status);
                    return;
                }
            }

            SimpleMailMessage message = new SimpleMailMessage();
            message.setTo(user.getEmail());
            message.setSubject(subject);
            message.setText(content);

            mailSender.send(message);
            log.info("[{}] 상태 변경 메일 전송 완료: {}", status, user.getEmail());
        } catch (Exception e) {
            log.error("이메일 전송 실패 (userId={}, email={}): {}", user.getId(), user.getEmail(), e.getMessage(), e);
        }
    }

    public void sendNewPassword(String email, String newPassword) {
        try {
            SimpleMailMessage message = new SimpleMailMessage();
            message.setTo(email);
            message.setSubject("[Dev-Station] 임시 비밀번호 안내");
            message.setText("""
                안녕하세요. Dev-Station 입니다.

                비밀번호 재설정 요청에 따라 임시 비밀번호를 발급해드렸습니다.
                아래의 비밀번호로 로그인 후, 반드시 새 비밀번호로 변경해주세요.

                임시 비밀번호: %s
                """.formatted(newPassword));

            mailSender.send(message);
            log.info("[비밀번호 재설정] 임시 비밀번호 전송 완료: {}", email);

        } catch (Exception e) {
            log.error("임시 비밀번호 전송 실패: {}", e.getMessage(), e);
            throw new ErrorException(ErrorCode.EMAIL_SEND_FAILED);
        }
    }
}

File: backend/src/main/java/com/backend/api/user/service/EmailService.kt
package com.backend.api.user.service

import com.backend.domain.user.entity.AccountStatus
import com.backend.domain.user.entity.User
import com.backend.domain.user.entity.VerificationCode
import com.backend.domain.user.repository.VerificationCodeRepository
import com.backend.domain.userPenalty.entity.UserPenalty
import com.backend.global.exception.ErrorCode
import com.backend.global.exception.ErrorException
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.reactive.function.BodyInserters
import org.springframework.web.reactive.function.client.WebClient
import java.security.SecureRandom
import java.time.LocalDateTime
import java.util.*

@Service
class EmailService(
    private val verificationCodeRepository: VerificationCodeRepository,

    @Value("\${mailgun.api-key}")
    private val mailgunApiKey: String,

    @Value("\${mailgun.domain}")
    private val mgDomain: String,

    @Value("\${mailgun.from}")
    private val fromEmail: String
) {

    private val log = LoggerFactory.getLogger(EmailService::class.java)

    private val CHAR_SET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    private val CODE_LENGTH = 6

    private val mailgunClient: WebClient = WebClient.builder()
        .baseUrl("https://api.mailgun.net/v3/$mgDomain")
        .defaultHeaders {
            val auth = "api:$mailgunApiKey"
            val encoded = Base64.getEncoder().encodeToString(auth.toByteArray())
            it.set("Authorization", "Basic $encoded")
        }
        .build()

    // 메일 발송
    private fun sendEmail(toEmail: String, subject: String, content: String) {
        try {
            val form = LinkedMultiValueMap<String, String>().apply {
                add("from", fromEmail)
                add("to", toEmail)
                add("subject", subject)
                add("text", content)
            }

            val response = mailgunClient.post()
                .uri("/messages")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .body(BodyInserters.fromFormData(form))
                .retrieve()
                .bodyToMono(String::class.java)
                .block()

            log.info("메일 전송 완료: {} | 응답={}", toEmail, response)

        } catch (e: Exception) {
            log.error("메일 전송 실패: {}", e.message)
            throw ErrorException(ErrorCode.EMAIL_SEND_FAILED)
        }
    }


    // 인증코드 생성 및 발송
    @Transactional
    fun createAndSendVerificationCode(email: String) {
        verificationCodeRepository.findByEmail(email)?.let {
            verificationCodeRepository.delete(it)
        }

        val code = generateVerificationCode()

        val verification = VerificationCode(
            email = email,
            code = code,
            expiresAt = LocalDateTime.now().plusMinutes(5),
            verified = false
        )

        verificationCodeRepository.save(verification)

        val content = """
            안녕하세요. Dev-Station 입니다.

            아래 인증 코드를 입력해 주세요.

            인증코드: $code
            (유효시간 5분)
        """.trimIndent()

        sendEmail(email, "[Dev-Station] 이메일 인증 코드", content)
    }


    private fun generateVerificationCode(): String {
        val random = SecureRandom()
        return buildString(CODE_LENGTH) {
            repeat(CODE_LENGTH) {
                append(CHAR_SET[random.nextInt(CHAR_SET.length)])
            }
        }
    }


    // 인증코드 검증
    fun verifyCode(email: String, code: String) {
        val verification = verificationCodeRepository.findByEmail(email)
            ?: throw ErrorException(ErrorCode.INVALID_VERIFICATION_CODE)

        if (verification.isExpired())
            throw ErrorException(ErrorCode.EXPIRED_VERIFICATION_CODE)

        if (verification.code != code)
            throw ErrorException(ErrorCode.INVALID_VERIFICATION_CODE)

        verification.markAsVerified()
        verificationCodeRepository.save(verification)

        log.info("[이메일 인증] 인증 성공: {}", email)
    }

    fun isVerified(email: String): Boolean =
        verificationCodeRepository.findByEmail(email)?.verified ?: false


    // 회원 가입 환영 메일
    fun sendWelcomeMail(user: User) {
        val content = """
            안녕하세요, ${user.name}님 👋

            Dev-Station 가입을 환영합니다!
        """.trimIndent()

        sendEmail(user.email, "[Dev-Station] 회원가입을 환영합니다!", content)
    }


    //계정 상태 변경 메일
    fun sendStatusChangeMail(user: User, penalty: UserPenalty?) {

        val (subject, content) = when (user.accountStatus) {

            AccountStatus.SUSPENDED -> {
                val reason = penalty?.reason ?: "사유 정보 없음"
                val endAt = penalty?.endAt ?: "미정"

                "[Dev-Station] 계정 일시정지 안내" to """
                    안녕하세요, ${user.name}님.

                    회원님의 계정이 일시정지되었습니다.
                    사유: $reason
                    종료일: $endAt
                """.trimIndent()
            }

            AccountStatus.BANNED -> {
                val reason = penalty?.reason ?: "사유 정보 없음"
                "[Dev-Station] 계정 영구정지 안내" to """
                    안녕하세요, ${user.name}님.

                    회원님의 계정이 영구 정지되었습니다.
                    사유: $reason
                """.trimIndent()
            }

            AccountStatus.ACTIVE -> {
                "[Dev-Station] 계정 복구 안내" to """
                    안녕하세요, ${user.name}님.

                    회원님의 계정이 정상으로 복구되었습니다.
                """.trimIndent()
            }

            AccountStatus.DEACTIVATED -> {
                "[Dev-Station] 탈퇴 완료 안내" to """
                    안녕하세요, ${user.name}님.

                    회원님의 계정 탈퇴 처리가 완료되었습니다.
                """.trimIndent()
            }
        }

        sendEmail(user.email, subject, content)
    }


    // 임시 비밀번호 발송
    fun sendNewPassword(email: String, newPassword: String) {

        val content = """
            안녕하세요. Dev-Station 입니다.

            요청하신 임시 비밀번호를 발급해드립니다.
            로그인 후 반드시 새 비밀번호로 변경해주세요.

            임시 비밀번호: $newPassword
        """.trimIndent()

        sendEmail(email, "[Dev-Station] 임시 비밀번호 안내", content)
    }
}
